staging/src/k8s.io/csi-translation-lib/plugins/in_tree_volume.go

```diff
@@ -124,38 +124,85 @@ func addTopology(pv *v1.PersistentVolume, topologyKey string, zones []string) er
 	}
 
 	// Make sure the necessary fields exist
-	pv.Spec.NodeAffinity = new(v1.VolumeNodeAffinity)
-	pv.Spec.NodeAffinity.Required = new(v1.NodeSelector)
-	pv.Spec.NodeAffinity.Required.NodeSelectorTerms = make([]v1.NodeSelectorTerm, 1)
+	if pv.Spec.NodeAffinity == nil {
+		pv.Spec.NodeAffinity = new(v1.VolumeNodeAffinity)
+	}
+
+	if pv.Spec.NodeAffinity.Required == nil {
+		pv.Spec.NodeAffinity.Required = new(v1.NodeSelector)
+	}
+
+	if len(pv.Spec.NodeAffinity.Required.NodeSelectorTerms) == 0 {
+		pv.Spec.NodeAffinity.Required.NodeSelectorTerms = make([]v1.NodeSelectorTerm, 1)
+	}
 
 	topology := v1.NodeSelectorRequirement{
 		Key:      topologyKey,
 		Operator: v1.NodeSelectorOpIn,
 		Values:   zones,
 	}
 
-	pv.Spec.NodeAffinity.Required.NodeSelectorTerms[0].MatchExpressions = append(
-		pv.Spec.NodeAffinity.Required.NodeSelectorTerms[0].MatchExpressions,
-		topology,
-	)
+	// add the CSI topology to each term
+	for i := range pv.Spec.NodeAffinity.Required.NodeSelectorTerms {
+		pv.Spec.NodeAffinity.Required.NodeSelectorTerms[i].MatchExpressions = append(
+			pv.Spec.NodeAffinity.Required.NodeSelectorTerms[i].MatchExpressions,
+			topology,
+		)
+	}
 
 	return nil
 }
 
-// translateTopology converts existing zone labels or in-tree topology to CSI topology.
+// removeTopology removes the topology from the given PV. Return false
+// if the topology key is not found
+func removeTopology(pv *v1.PersistentVolume, topologyKey string) bool {
+	// Make sure the necessary fields exist
+	if pv == nil || pv.Spec.NodeAffinity == nil || pv.Spec.NodeAffinity.Required == nil ||
+		pv.Spec.NodeAffinity.Required.NodeSelectorTerms == nil || len(pv.Spec.NodeAffinity.Required.NodeSelectorTerms) == 0 {
+		return false
+	}
+
+	succeed := false
+	for termIndex, nodeSelectorTerms := range pv.Spec.NodeAffinity.Required.NodeSelectorTerms {
+		nsrequirements := nodeSelectorTerms.MatchExpressions
+
+		index := -1
+		for i, nodeSelectorRequirement := range nsrequirements {
+			if nodeSelectorRequirement.Key == topologyKey {
+				index = i
+				break
+			}
+		}
+		// We found the key that need to be removed
+		if index != -1 {
+			nsrequirements[len(nsrequirements)-1], nsrequirements[index] = nsrequirements[index], nsrequirements[len(nsrequirements)-1]
+			pv.Spec.NodeAffinity.Required.NodeSelectorTerms[termIndex].MatchExpressions = nsrequirements[:len(nsrequirements)-1]
+			succeed = true
+		}
+	}
+
+	return succeed
+}
+
+// translateTopologyFromInTreeToCSI converts existing zone labels or in-tree topology to CSI topology.
 // In-tree topology has precedence over zone labels.
-func translateTopology(pv *v1.PersistentVolume, topologyKey string) error {
```

Comment:
btw do you know come this function only translated zone and not region? it looks like azuredd and cinder provisioners do set region, so the info gets lost on translation, potentially it is a bug for them.

OTOH, your new translateTopologyFromCSIToInTree does handle region makes more sense to me.

Reply:
Hmm, I am not familiar with those two so I could not tell if this is going to be a problem. But I guess their driver has its own CSI migration test and if those tests are working fine. Then it should be good?

Reply:
Yeah let's leave it as is, I may open a separate issue if necessary but I was just curious since now there is some asymmetry between translateTopologyFromCSIToInTree and translateTopologyFromInTreeToCSI.

Reply:
cc @andyzhangx @gnufied for region questions on azure and openstack.

At least in aws and gcp cases, you can derive the region from the zone.